sklearn/metrics/regression.py

```diff
@@ -241,26 +317,49 @@ def explained_variance_score(y_true, y_pred, sample_weight=None):
     >>> y_pred = [2.5, 0.0, 2, 8]
     >>> explained_variance_score(y_true, y_pred)  # doctest: +ELLIPSIS
     0.957...
+    >>> y_true = [[0.5, 1], [-1, 1], [7, -6]]
+    >>> y_pred = [[0, 2], [-1, 2], [8, -5]]
+    >>> explained_variance_score(y_true, y_pred, output_weights='uniform')  # doctest: +ELLIPSIS
+    0.983...
 
     """
-    y_type, y_true, y_pred = _check_reg_targets(y_true, y_pred)
+    y_type, y_true, y_pred, output_weights = _check_reg_targets(
+        y_true, y_pred, output_weights)
 
-    if y_type != "continuous":
+    if y_type not in ["continuous", "continuous-multioutput"]:
         raise ValueError("{0} is not supported".format(y_type))
 
-    _, numerator = _average_and_variance(y_true - y_pred, sample_weight)
-    _, denominator = _average_and_variance(y_true, sample_weight)
-    if denominator == 0.0:
-        if numerator == 0.0:
-            return 1.0
-        else:
-            # arbitrary set to zero to avoid -inf scores, having a constant
-            # y_true is not interesting for scoring a regression anyway
-            return 0.0
-    return 1 - numerator / denominator
-
-
-def r2_score(y_true, y_pred, sample_weight=None):
+    y_diff_avg = np.average(y_true - y_pred, weights=sample_weight, axis=0)
+    numerator = np.average((y_true - y_pred - y_diff_avg) ** 2,
+                           weights=sample_weight, axis=0)
+
+    y_true_avg = np.average(y_true, weights=sample_weight, axis=0)
+    denominator = np.average((y_true - y_true_avg) ** 2,
+                             weights=sample_weight, axis=0)
+
+    nonzero_numerator = numerator != 0
+    nonzero_denominator = denominator != 0
+    valid_score = nonzero_numerator & nonzero_denominator
+    output_scores = np.ones(y_true.shape[1])
+
+    output_scores[valid_score] = 1 - (numerator[valid_score] /
+                                      denominator[valid_score])
+    output_scores[nonzero_numerator & ~nonzero_denominator] = 0.
+    if output_weights is None:
+        # return scores individually
+        return output_scores
+    elif output_weights == 'uniform':
+        # passing None as weights results is uniform mean
+        output_weights = None
+    elif output_weights == 'variance':
+        output_weights = denominator
+
+    return np.average(output_scores, weights=output_weights)
+
+
+def r2_score(y_true, y_pred,
+             output_weights='uniform',
```

Comment:
Regression with respect to the single target behavior or regression with respect to a previous version of this PR?